sdk/core/azure-core/inc/credentials/credentials.hpp

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+namespace azure
+{
+namespace core
+{
+namespace credentials
+{
+
+class Credential
+{
+public:
+  class _internal;
+  virtual ~Credential() noexcept;
+
+protected:
+  Credential();
+
+  Credential(Credential const& other);
+  Credential& operator=(Credential const& other);
+
+private:
+  virtual void SetScopes(std::string const& scopes);
+};
+
+class TokenCredential : public Credential
+{
+public:
+  class _internal;
+  ~TokenCredential() noexcept override;
+
+protected:
+  TokenCredential();
+
+  TokenCredential(TokenCredential const& other);
+  TokenCredential& operator=(TokenCredential const& other);
+
+private:
+  struct Token;
+  std::shared_ptr<Token> _token;
```

Comment:
I'm not really seeing the shared_ptr buying you anything here (other than more memory allocations)

Reply:
I think it buys me thread safety when updating, atomicity of updating TokenString+Expiration+Scopes.
So you can be sure when you get Token, and then read Expiration, you know that it corresponds to the TokenString, and TokenString won't update.
If another thread updated token after that, you can call GetToken() again, and it will also be consistent.

Reply:
Assigning over a shared_ptr is *not* thread safe. It buys you no thread safety here.

Reply:
(The thread safety guarantee `shared_ptr` provides is that if more than one `shared_ptr` is bound to a given reference count, modifications **of the reference count, and only of the reference count** are protected from data races. Concurrent modifications of the same `shared_ptr` instances, or concurrent modifications of the `T` to which the `shared_ptr<T>` points are data races.)

Reply:
You may want `std::atomic<std::shared_ptr<T>>`, but it has a mutex internally, also it's not available until C++20.

Reply:
I used std::mutex.